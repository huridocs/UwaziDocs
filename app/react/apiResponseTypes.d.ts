import { UserGroupSchema } from 'shared/types/userGroupType';
import { UserSchema } from 'shared/types/userType';
import {
  PreserveConfig,
  Settings,
  SettingsFilterSchema,
  SettingsLinkSchema,
} from 'shared/types/settingsType';
import { LanguageSchema } from 'shared/types/commonTypes';

export interface GroupMemberSchema {
  refId: string;
  username: string;
}

export interface ClientUserGroupSchema extends Omit<UserGroupSchema, '_id' | 'members'> {
  _id?: string;
  members: GroupMemberSchema[];
}

export interface ClientUserSchema extends Omit<UserSchema, '_id' | 'groups'> {
  _id?: string;
  groups?: {
    _id: string;
    name: string;
  }[];
}

export interface ClientSettingsFilterSchema extends Omit<SettingsFilterSchema, '_id'> {
  _id?: string;
}

export interface ClientSublink extends Omit<ClientSettingsLinkSchema, 'sublinks'> {}

export interface ClientSettingsLinkSchema extends Omit<SettingsLinkSchema, '_id' | 'sublinks'> {
  _id?: string;
  sublinks?: ClientSublink[];
}

export interface ClientPreserveConfig extends Omit<PreserveConfig, 'config'> {
  config: {
    token: string;
    template: string;
    user?: string;
  }[];
}

export interface ClientLanguageSchema extends Omit<LanguageSchema, '_id'> {
  _id?: string;
}

export interface ClientSettings
  extends Omit<Settings, '_id | filters | links | features | languages'> {
  _id?: string;
  filters?: ClientSettingsFilterSchema[];
  languages?: ClientLanguageSchema[];
  links?: ClientSettingsLinkSchema[];
  features?: Omit<Settings['features'], 'preserve | metadataExtraction'> & {
    preserve?: ClientPreserveConfig;
    metadataExtraction?: {
      url: string;
      templates?: {
        template: string;
        properties: string[];
      }[];
    };
    [k: string]: unknown | undefined;
  };
}
