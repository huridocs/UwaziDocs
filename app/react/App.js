import React from 'react';
import { Provider } from 'react-redux';
import { routes } from './Routes';
import CustomProvider from './App/Provider';
import { store } from './store';

const App = () => (
  <Provider store={store}>
    <CustomProvider>{routes}</CustomProvider>
  </Provider>
);

export default App;
