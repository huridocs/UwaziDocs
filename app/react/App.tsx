import React from 'react';
import { RouterProvider, createBrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { getRoutes } from './Routes';
import CustomProvider from './App/Provider';
import { store } from './store';

const reduxState = store?.getState();

const router = createBrowserRouter(
  getRoutes(reduxState?.settings.collection.toJS(), reduxState?.user.get('_id'))
);

const App = () => (
  <Provider store={store as any}>
    <CustomProvider>
      <RouterProvider router={router} fallbackElement={null} />
    </CustomProvider>
  </Provider>
);

export { App };
