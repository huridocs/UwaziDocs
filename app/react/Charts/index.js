import loadable from '@loadable/component';

import colorScheme from './utils/colorScheme';
import arrayUtils from './utils/arrayUtils';
import ExtendedTooltip from './components/ExtendedTooltip';

const Bar = loadable(
  async () => import(/* webpackChunkName: "LazyLoadBar" */ './components/Bar.js')
);

const ColoredBar = loadable(
  async () => import(/* webpackChunkName: "LazyLoadColoredBar" */ './components/ColoredBar.js')
);

const Pie = loadable(
  async () => import(/* webpackChunkName: "LazyLoadPie" */ './components/Pie.js')
);

const StackedDualBarChart = loadable(
  async () =>
    import(
      /* webpackChunkName: "LazyLoadStackedDualBarChart" */ './components/StackedDualBarChart.js'
    )
);

export { Bar, ColoredBar, ExtendedTooltip, Pie, StackedDualBarChart, colorScheme, arrayUtils };
