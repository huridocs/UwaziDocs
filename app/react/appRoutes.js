import { getRoutes } from './Routes';
import { settingsAtom, atomStore, userAtom } from './V2/atoms';

export const routes = getRoutes && getRoutes(atomStore.get(settingsAtom), atomStore.get(userAtom)?._id);

