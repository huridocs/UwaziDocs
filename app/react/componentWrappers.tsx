/* eslint-disable comma-spacing */
/* eslint-disable react/no-multi-comp */
import React, { useContext } from 'react';
import {
  useLocation,
  useMatches,
  useNavigate,
  useOutlet,
  useParams,
  useSearchParams,
} from 'react-router-dom';
import { AppMainContext } from './App/AppMainContext';

const withRouter =
  <T,>(Component: React.ComponentClass<T & { params?: any }, any>) =>
  (props: T & { params?: any }) => {
    const location = useLocation();
    const navigate = useNavigate();
    const params = useParams();
    const matches = useMatches();
    const [searchParams, setSearchParams] = useSearchParams();
    return (
      <Component
        {...props}
        navigate={navigate}
        location={location}
        params={{ ...params, ...(props.params ? props.params : {}) }}
        matches={matches}
        searchParams={searchParams}
        setSearchParams={setSearchParams}
      />
    );
  };

const withContext =
  <T,>(Component: React.ComponentClass<T, any>) =>
  (props: T) => {
    const mainContext = useContext(AppMainContext);
    return <Component {...props} mainContext={mainContext} />;
  };

const withOutlet =
  <T,>(Component: React.ComponentClass<T, any>) =>
  (props: T) => {
    const outlet = useOutlet(AppMainContext);
    return <Component {...props} outlet={outlet} />;
  };
export { withRouter, withContext, withOutlet };
