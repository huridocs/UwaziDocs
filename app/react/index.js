import React from 'react';
import ReactDOM from 'react-dom';

import * as Sentry from '@sentry/react';
import { BrowserTracing } from '@sentry/tracing';

import { AppContainer } from 'react-hot-loader';

import { SENTRY_APP_DSN } from 'app/config';
import App from './App';

import './App/sockets';

if (SENTRY_APP_DSN) {
  Sentry.init({
    dsn: SENTRY_APP_DSN,
    integrations: [new BrowserTracing()],
    tracesSampleRate: 0.1,
  });
}

const render = Component => {
  ReactDOM.hydrate(
    <AppContainer>
      <Component />
    </AppContainer>,
    document.getElementById('root')
  );
};

render(App);

if (module.hot) {
  module.hot.accept('./App', () => {
    const nextApp = require('./App');
    render(nextApp);
  });
}
