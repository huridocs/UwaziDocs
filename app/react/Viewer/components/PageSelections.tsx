import React from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { Highlight } from 'react-text-selection-handler';
import { ClientFile, IStore } from 'app/istore';
import { isBlobFile } from 'shared/tsUtils';
import { IImmutable } from 'shared/types/Immutable';
import { SelectionRectanglesSchema } from 'shared/types/commonTypes';

const mapStateToProps = (state: IStore) => {
  const currentDocumentId = state.documentViewer.doc.get('defaultDoc')?.get('_id');
  const entityDocuments = state.documentViewer.doc
    .get('documents')
    ?.filter(document => !isBlobFile(document)) as unknown as IImmutable<ClientFile[]>;
  const entityDocument = entityDocuments.filter(
    document => document?.get('_id') === currentDocumentId
  );

  return {
    newSelections: state.documentViewer.metadataExtraction.get('selections'),
    entityDocument: entityDocument.get(0),
  };
};

const connector = connect(mapStateToProps);

type mappedProps = ConnectedProps<typeof connector>;

const PageSelectionsComponent = ({ newSelections, entityDocument }: mappedProps) => {
  const currentSelections = entityDocument.toJS().extractedMetadata?.length
    ? entityDocument
        .toJS()
        .extractedMetadata!.map(currentSelection => ({ ...currentSelection, isCurrent: true }))
    : [];

  const selections = [...currentSelections, ...newSelections.toJS()];

  if (selections.length) {
    return (
      <>
        {selections.map(selection => {
          const selected = selection.selection;
          const rectangles = (selected.selectionRectangles as SelectionRectanglesSchema).map(
            rectangle => ({
              regionId: rectangle.page,
              ...rectangle,
            })
          );
          const highlight = { text: selected.text, selectionRectangles: rectangles };
          return (
            <Highlight
              textSelection={highlight}
              color={selection.isCurrent ? '#359990' : '#5cb85c'}
            />
          );
        })}
      </>
    );
  }

  return <></>;
};

const container = connector(PageSelectionsComponent);

export { container as PageSelections };
