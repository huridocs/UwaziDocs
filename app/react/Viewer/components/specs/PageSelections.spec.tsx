/**
 * @jest-environment jsdom
 */
import React from 'react';
import Immutable from 'immutable';
import { RenderResult } from '@testing-library/react';
import { defaultState, renderConnectedContainer } from 'app/utils/test/renderConnected';
import { ClientEntitySchema, ClientFile, ExtractedMetadataSelection } from 'app/istore';
import { PageSelections } from '../PageSelections';

const defaultEntityDocument: ClientFile = {
  _id: '62f52bddc6897a159347cf6b',
  originalname: 'Get_Started_With_Smallpdf.pdf',
  type: 'document',
  totalPages: 2,
  entity: 'bc3prdymqj',
  filename: '1660234717101q3vq1v0vo7h.pdf',
  language: 'eng',
  extractedMetadata: [
    {
      propertyID: '62f290a54dd69a2472936453',
      name: 'my_property',
      selection: {
        text: 'Selected text',
        selectionRectangles: [
          {
            top: 91,
            left: 32,
            width: 30,
            height: 26,
            page: '2',
          },
        ],
      },
    },
    {
      name: 'title',
      selection: {
        text: 'The title of the PDF',
        selectionRectangles: [
          {
            top: 35,
            left: 261,
            width: 272,
            height: 41,
            page: '1',
          },
        ],
      },
    },
  ],
};

describe('Page selections highlights', () => {
  let renderResult: RenderResult;
  let entity: ClientEntitySchema = {
    _id: '62f52bdcc6897a159347cf59',
  };
  let file: any | ClientFile;
  let selections: ExtractedMetadataSelection[];

  beforeEach(() => {
    file = defaultEntityDocument;
    selections = [];
  });

  const render = () => {
    const store = {
      ...defaultState,
      documentViewer: {
        doc: Immutable.fromJS({
          defaultDoc: file,
        }),
        sidepanel: {
          metadata: entity,
        },
        metadataExtraction: Immutable.fromJS({
          selections,
        }),
      },
    };
    ({ renderResult } = renderConnectedContainer(<PageSelections />, () => store));
  };

  it('should only render when editing the entity and has a document', () => {
    entity = {};
    render();
    expect(renderResult.container.innerHTML).toBe('');

    entity = { _id: 'some_id' };
    file = {};
    render();
    expect(renderResult.container.innerHTML).toBe('');
  });

  it('should highlight existing selections', () => {
    render();
    expect(renderResult.container.children.length).toBe(2);
  });

  it('should highligh new selections', () => {
    selections = [
      {
        propertyID: '4356fdsassda',
        name: 'property_name',
        timestamp: 'today',
        selection: { text: 'new selected text', selectionRectangles: [{ top: 10, page: '1' }] },
      },
    ];
    render();
    expect(renderResult.container.children.length).toBe(3);
  });
});
