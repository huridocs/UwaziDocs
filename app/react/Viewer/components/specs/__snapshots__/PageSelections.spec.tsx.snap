// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page selections highlights should adjust selections by the pdf scaling factor 1`] = `
<body>
  <div>
    <div
      class="selection"
      data-testid="oldProperty"
    >
      <div
        class="highlight-rectangle"
        style="top: 136.5px; left: 48px; width: 45px; height: 39px; padding: 0px; margin: 0px; position: absolute; display: block; mix-blend-mode: darken; opacity: 0.5; background-color: rgb(177, 247, 163); z-index: 1;"
      />
    </div>
    <div
      class="selection"
      data-testid="oldTitle"
    >
      <div
        class="highlight-rectangle"
        style="top: 52.5px; left: 391.5px; width: 408px; height: 61.5px; padding: 0px; margin: 0px; position: absolute; display: block; mix-blend-mode: darken; opacity: 0.5; background-color: rgb(177, 247, 163); z-index: 1;"
      />
    </div>
  </div>
</body>
`;
